File: src/lexer/regex/RegexParser.cpp
module;

export module language.tokenizer:regexparser;

import <iostream>;
import <map>;
import <set>;
import <stack>;

import language;

import :dfa;
import :fsm;

import utility;

export enum class regexNodeType { CHARS, ALTERNATION, STAR, OPTION, STRING };
export enum class regexOperatorType { STRING, ALTERNATION, PAREN };

using tree_t = AbstractSyntaxTree<regexNodeType>;
using node_t = tree_t::Node;
using internal_node_t = tree_t::InternalNode;
using leaf_node_t = tree_t::LeafNode;

export class RegexParser {
	RegexParser () = default;

	std::stack<node_t::pointer_t> stack {};
	std::stack<regexOperatorType> operator_stack;

	bool hasOperatorsLeft ();

	node_t::pointer_t stackPop ();

	regexOperatorType operatorStackPop ();

	regexOperatorType operatorStackPeek ();

	static node_t::pointer_t& append_element (node_t::pointer_t& x, node_t::pointer_t y);

	void chars (const std::set<char>& chars);

	void concat ();

	void alternation ();

	void begin_group ();

	void end_group ();

	void apply (regexNodeType op);

	tree_t result ();

public:
	static tree_t parse (const std::string& regex);
};

std::ostream& operator<< (std::ostream& os, regexNodeType t);

std::ostream& operator<< (std::ostream& os, const node_t::pointer_t& node);

bool RegexParser::hasOperatorsLeft () {
	return !operator_stack.empty();
}

node_t::pointer_t RegexParser::stackPop () {
	node_t::pointer_t res = std::move(stack.top());
	stack.pop();
	return res;
}

regexOperatorType RegexParser::operatorStackPeek () {
	return operator_stack.top();
}

node_t::pointer_t& RegexParser::append_element (node_t::pointer_t& x, node_t::pointer_t y) {
	x->asInternalNode()->add_child(std::move(y));
	return x;
}

void RegexParser::chars (const std::set<char>& chars) {
	stack.push(tree_t::createLeafNode(regexNodeType::CHARS, chars));
}

void RegexParser::concat () {
	while (hasOperatorsLeft() && operatorStackPeek() == regexOperatorType::STRING) {
		node_t::pointer_t x2 = stackPop();
		node_t::pointer_t x1 = stackPop();
		operatorStackPop();
		stack.push(std::move(append_element(x1, std::move(x2))));
	}
	operator_stack.push(regexOperatorType::STRING);
}

void RegexParser::alternation () {
	while (hasOperatorsLeft() && operatorStackPeek() != regexOperatorType::PAREN) {
		node_t::pointer_t x2 = stackPop();
		node_t::pointer_t x1 = stackPop();
		switch (operatorStackPop()) {
			case regexOperatorType::ALTERNATION:
				if (x1->nodeType != regexNodeType::ALTERNATION)
					stack.push(tree_t::createInternalNode(regexNodeType::ALTERNATION, std::move(x1), std::move(x2)));
				else {
					x1->asInternalNode()->add_child(std::move(x2));
					stack.push(std::move(x1));
				}
				break;
			case regexOperatorType::STRING:
				x1->asInternalNode()->add_child(std::move(x2));
				stack.push(std::move(x1));
				break;
			default:
				throw std::runtime_error("Error parsing regex: unknown operator on operator stack");
		}
	}
	operator_stack.push(regexOperatorType::ALTERNATION);
	stack.push(tree_t::createInternalNode(regexNodeType::STRING, {}));
}

void RegexParser::begin_group () {
	operator_stack.push(regexOperatorType::PAREN);
	stack.push(tree_t::createInternalNode(regexNodeType::STRING, {}));
}

void RegexParser::end_group () {
	for (regexOperatorType t; (t = operatorStackPop()) != regexOperatorType::PAREN;) {
		node_t::pointer_t x2 = stackPop();
		node_t::pointer_t x1 = stackPop();
		switch (t) {
			case regexOperatorType::ALTERNATION:
				if (x1->nodeType != regexNodeType::ALTERNATION) {
					stack.push(tree_t::createInternalNode(regexNodeType::ALTERNATION, std::move(x1), std::move(x2)));
				} else {
					x1->asInternalNode()->add_child(std::move(x2));
					stack.push(std::move(x1));
				}
				break;
			case regexOperatorType::STRING:
				x1->asInternalNode()->add_child(std::move(x2));
				stack.push(std::move(x1));
				break;
			default:
				throw std::runtime_error("Error parsing regex: unknown operator on operator stack");
		}
	}
}

void RegexParser::apply (regexNodeType op) {
	stack.push(tree_t::createInternalNode(op, stackPop()));
}

tree_t RegexParser::result () {
	return tree_t { std::move(stack.top()) };
}

tree_t RegexParser::parse (const std::string& regex) {
	RegexParser reParser;

	reParser.begin_group();
	for (int i = 0; i < regex.size(); i++) {
		char c = regex[i];
		switch (c) {
			case '|':
				reParser.alternation();
				break;
			case '(':
				reParser.concat();
				reParser.begin_group();
				break;
			case ')':
				reParser.end_group();
				break;
			case '*':
				reParser.apply(regexNodeType::STAR);
				break;
			case '?':
				reParser.apply(regexNodeType::OPTION);
				break;
			case '.':
				reParser.chars(DFA::all_chars());
				break;
			case '[':
				reParser.concat();
				{
					bool conjugate = false;
					std::set<char> chars;
					if (regex[i + 1] == '^') {
						conjugate = true;
						chars = FSM::all_chars();
						i++;
					}
					while ((c = regex[++i]) != ']') {
						if (c == '\\') {
							switch (regex[++i]) {
								case 'w':
									for (const char ch: FSM::word_chars()) {
										if (conjugate) {
											chars.erase(ch);
										} else {
											chars.emplace(ch);
										}
									}
									break;
								case 'l':
									for (const char ch: FSM::letters()) {
										if (conjugate) {
											chars.erase(ch);
										} else {
											chars.emplace(ch);
										}
									}
									break;
								case 'd':
									for (const char ch: FSM::digits()) {
										if (conjugate) {
											chars.erase(ch);
										} else {
											chars.emplace(ch);
										}
									}
									break;
								case 'n':
									if (conjugate) {
										chars.erase('\n');
									} else {
										chars.emplace('\n');
									}
									break;
								case 't':
									if (conjugate) {
										chars.erase('\t');
									} else {
										chars.emplace('\t');
									}
									break;
								default:
									if (conjugate) {
										chars.erase(regex[i]);
									} else {
										chars.emplace(regex[i]);
									}
							}
						} else if (conjugate) {
							chars.erase(c);
						} else {
							chars.emplace(c);
						}
					}
					if (chars.empty()) throw std::runtime_error("Error parsing regex: empty character class");
					reParser.chars(chars);
				}
				break;
			case ']':
				throw std::runtime_error("Error parsing regex: mismatched brackets []");
			case '\\':
				c = regex[++i];
				switch (c) {
					case 'w':
						reParser.concat();
						reParser.chars(FSM::word_chars());
						break;
					case 'l':
						reParser.concat();
						reParser.chars(FSM::letters());
						break;
					case 'd':
						reParser.concat();
						reParser.chars(FSM::digits());
						break;
					case 'n':
						reParser.concat();
						reParser.chars({ '\n' });
						break;
					case 't':
						reParser.concat();
						reParser.chars({ '\t' });
						break;
					default:
						reParser.concat();
						reParser.chars({ c });
				}
				break;
			default:
				reParser.concat();
				reParser.chars({ c });
				break;
		}
	}
	reParser.end_group();
	return reParser.result();
}

regexOperatorType RegexParser::operatorStackPop () {
	auto res = operator_stack.top();
	operator_stack.pop();
	return res;
}

std::ostream& operator<< (std::ostream& os, regexNodeType t) {
	switch (t) {
		case regexNodeType::CHARS:
			return os << "chars";
		case regexNodeType::ALTERNATION:
			return os << "<|>";
		case regexNodeType::STAR:
			return os << "<*>";
		case regexNodeType::OPTION:
			return os << "<?>";
		case regexNodeType::STRING:
			return os << "<#>";
	}
	return os;
}

// TODO: generify to apply to any parse tree and move to AbstractSyntaxTree.ixx

std::ostream& operator<< (std::ostream& os, const node_t::pointer_t& node) {
	int indent_level = 0;
	std::stack<const node_t::pointer_t*> stk;
	stk.push(&node);

	while (!stk.empty()) {
		const node_t::pointer_t& currentNode = *stk.top();
		stk.pop();

		if (!currentNode) {
			indent_level--;
			for (int i = 0; i < indent_level; i++) os << "    ";
			os << "}\n";
			continue;
		}

		for (int i = 0; i < indent_level; i++) os << "    ";
		os << currentNode->nodeType;

		if (currentNode->isLeaf()) {
			const leaf_node_t* leaf = currentNode->asLeafNode();

			const auto& s = leaf->get_value<std::set<char>>();

			os << "[" << std::string { s.begin(), s.end() } << "]" << "\n";
		} else {
			const internal_node_t* as_internal = currentNode->asInternalNode();
			os << " {\n";
			stk.push({});
			typeof(stk) reverse;
			for (const node_t::pointer_t& child: *as_internal) {
				reverse.push(&child);
			}
			while (!reverse.empty()) {
				stk.push(reverse.top());
				reverse.pop();
			}
			indent_level++;
		}
	}
	return os;
}

File: src/lexer/regex/RegexParser.js
import Regex, {
	AlternationNode,
	ConcatenationNode,
	KleeneStarNode,
	OptionalNode,
	LeafNode, KleenePlusNode
} from "./Regex.js";
import {ALL_CHARS, DIGIT_CHARS, LETTER_CHARS, SPACE_CHARS, WORD_CHARS} from "../fsm/FiniteStateMachine.js";

/**
 * @typedef {"STRING" | "ALTERNATION" | "PAREN"} RegexOperatorType
 */

class RegexParser {
	/** @type {Regex[]} */
    #stack = [];
    /** @type {RegexOperatorType[]} */
    #operatorStack = [];

    get hasOperatorsLeft() {
        return !!this.#stack.length;
    }

    /** @param {Set<string>} chars */
	chars (chars){
        this.#stack.push(new LeafNode(chars))
    }

	concat () {
        while (this.hasOperatorsLeft && this.#operatorStack[this.#operatorStack.length - 1] === "STRING") {
            const x2 = this.#stack.pop();
            const x1 = this.#stack.pop();
            this.#operatorStack.pop();

            this.#stack.push(x1.addChild(x2));
        }
        this.#operatorStack.push("STRING");
    }

	alternation () {
        while (this.hasOperatorsLeft && this.#operatorStack[this.#operatorStack.length - 1] !== "PAREN") {
            const x2 = this.#stack.pop();
            const x1 = this.#stack.pop();

            switch(this.#operatorStack.pop()){
                case "ALTERNATION":
                    if (!(x1 instanceof AlternationNode)) {
                        this.#stack.push(new AlternationNode(x1, x2));
                        break;
                    }
                case "STRING":
                    this.#stack.push(x1.addChild(x2));
                    break;

                default: throw new Error('Something went wrong :(');
            }
        }

        this.#operatorStack.push("ALTERNATION");
        this.#stack.push(new ConcatenationNode);
    }

	begin_group () {
        this.#operatorStack.push("PAREN");
        this.#stack.push(new ConcatenationNode);
    }

	end_group () {
        for (/** @type {RegexOperatorType} */ let t; (t = this.#operatorStack.pop()) !== "PAREN";) {
            const x2 = this.#stack.pop();
            const x1 = this.#stack.pop();

            switch (t) {
                case "ALTERNATION":
                    if (!(x1 instanceof AlternationNode)) {
                        this.#stack.push(new AlternationNode(x1, x2));
                        break;
                    }
                case "STRING":
                    this.#stack.push(x1.addChild(x2));
                    break;

                default:
                    throw new Error('Something went wrong :(');
            }
        }
    }

    /**
     * @param {function(Regex):Regex} op
     */
	apply (op){
        this.#stack.push(new op(this.#stack.pop()));
    }

    get result() {
        return this.#stack[0];
    }
}

/**
 * @param {string} escapedCharacter
 * @returns {string[]}
 */
function charsForEscape(escapedCharacter) {
	switch (escapedCharacter) {
		case 'w':
			return [...WORD_CHARS];
		case 'l':
			return [...LETTER_CHARS];
		case 'd':
			return [...DIGIT_CHARS];
		case 's':
			return [...SPACE_CHARS];
		case 'n':
			return ['\n']
		case 't':
			return ['\t'];
		default:
			return [escapedCharacter];
	}
}

/**
 * Parses a string into a {@link src}, which can be compiled into an {@link NFA}
 * @param {string} src
 * @returns {Regex}
 */
export function parseRegex(src) {
	const reParser = new RegexParser();

	reParser.begin_group();
	for (let i = 0; i < src.length; i++) {
		let c = src[i];
		switch (c) {
			case '|':
				reParser.alternation();
				break;
			case '(':
				reParser.concat();
				reParser.begin_group();
				break;
			case ')':
				reParser.end_group();
				break;
			case '+':
				reParser.apply(KleenePlusNode);
				break;
			case '*':
				reParser.apply(KleeneStarNode);
				break;
			case '?':
				reParser.apply(OptionalNode);
				break;
			case '.':
				reParser.concat();
				reParser.chars(ALL_CHARS);
				break;
			case '[':
				let conjugate = false;
				/** @type {Set<string>} */
				let chars = new Set;
				if (src[i + 1] === '^') {
					conjugate = true;
					chars = new Set(ALL_CHARS);
					i++;
				}

				while ((c = src[++i]) !== ']') {
					const addedLetters = c === '\\' ?
						charsForEscape(src[++i]) :
						[c];

					for (const ch of addedLetters) {
						if (conjugate) chars.delete(ch);
						else chars.add(ch);
					}
				}

				if (chars.size === 0) throw new Error("Error parsing regex: empty character class");

				reParser.concat();
				reParser.chars(chars);
				break;
			case ']':
				throw new Error("Error parsing regex: mismatched brackets []");
			case '\\':
				reParser.concat();
				reParser.chars(new Set(charsForEscape(src[++i])));
				break;

			default:
				reParser.concat();
				reParser.chars(new Set(c));
				break;
		}
	}
	reParser.end_group();
	return reParser.result;
}